### PR TITLE
Module-ize server routes and add static file support

### LINKED ISSUE #1148: Route to static  files in the --server mode
Most of my notes needs an attachment feature (mostly huge pdfs). I found no good way for this.

So I'm just added an new [static route](https://github.com/peko/TiddlyWiki5/blob/master/core/modules/commands/server.js#L227-L240) for "files" subfolder. 

It is ok to do it like this, or there is a better way / or plugin?

### PR DESCRIPTION
**Pull request by sukima**
Fixes #1148

Refactors route into module types to be loaded dynamically.
Adds a static asset route plugin (which leverages the new module type).

### CODE REVIEW

File: core/modules/serverroutes/get-tiddlers.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/*\
+title: $:/core/modules/serverroutes/get-tiddlers.js
```
Comment (Jermolene) [inline core/modules/serverroutes/get-tiddlers.js]: I think that this route actually gets a single tiddler, so should be singular not plural.

File: core/modules/serverroutes/put-tiddlers.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/*\
+title: $:/core/modules/serverroutes/put-tiddlers.js
```
Comment (Jermolene) [inline core/modules/serverroutes/put-tiddlers.js]: Similarly, I think this should named in the singular

**Comment by sukima** [comment]
@Jermolene Fixed based on your feed back and rebased.

**Comment by sukima** [comment]
I think this might also fix issue #1000

**Comment by Jermolene** [comment]
Hi @sukima thanks for this, and apologies for the delayed response.

Generally, I think the client-server architecture needs some love. The functionality was implemented to the point where it basically worked in common situations, but was then pretty much frozen. As part of addressing the specific missing functionality like static assets, there's some deeper things I think we should fix. So, some of what follows is critiques of your implementation, and some is suggestions of more ambitious approaches.

I'd like to explore reusing the route handling mechanism across three places: 

* handling incoming HTTP requests
* handling incoming `window.postMessage` requests in the browser. The current implementation of the plugin library tunnels HTTP-like requests over `window.postMessage`, which would also be useful for TiddlyDesktop-like things wanting to endow a file-based TiddlyWiki with an HTTP interface
* generating static files (which I'd like to be able to do both in the browser via zip and under Node.js)

Thus, I could define my website structure once, and then choose whether to serve exactly the same thing via HTTP, or by exporting static files. I'm interested in exposing the routes as simple declarative tiddlers so that we can provide a UI for users to easily add or review them.

Internally, the rough idea would be to implement objects with interfaces like `server`, `request` and `response` for the `window.postMessage` and static file cases.

Eventually, the HTTP handling needs to be able to use a different wiki for the client than that being used for the server. That would get rid of the current situation of having to have the same configuration on the server and in the browser. Thus, server routes should be given both the wiki object that they are hosted in (where they'd look for other modules etc) and the wiki object that is being served.

On to some specific points about this pull request.

Firstly, I think ideally the static assets plugin would should configure a core route, rather than containing the code for a route itself. That way people could add another set of static routes to eg their videos with another plugin, without needing to write the JS of a route.

Secondly, I think we should explore integrating with the TiddlyWiki external tiddler mechanism: the client wiki should get a tiddler with a `_canonical_uri` field pointing to the external asset. That implies that the `assets` folder is actually a tiddler folder, but with the special behaviour of being served externally by default.

Finally, what you've done so far is a perfectly good first step for these sort of enhancements, but I think it may make sense to set up a branch to hold these server enhancements as I think it will take a few iterations to get things right.

**Comment by sukima** [comment]
@Jermolene Ok, though I am not sure what the next actions would be. Seems you have a great deal more context on this then I. Perhaps this new vision could be articulated in a RFC like document?

As for a separate branch you have the control of that since GitHub only allows team contributors to make new branches. To get this started based on this Pull Request you could:

```console
$ git fetch origin pull/2502/head:server-v2
$ git checkout server-v2
$ git push origin server-v2
```

([Checking out pull requests locally](https://help.github.com/articles/checking-out-pull-requests-locally/))

**Comment by Jermolene** [comment]
Hi @sukima I'm only on mobile devices for the next few days, so I can't run the command line. I've created a new "http-server" branch, perhaps you could kindly somehow redirect the PR to it?

In terms of next steps, I won't be in a position to draft an RFC for a few days, but please feel free to ask any questions here, or on the new PR.

**Comment by sukima** [comment]
Moved to PR #2510 so new PR can be based on branch `http-server` instead of `master`

**Pull request closed without merging**